bool ready
bool from_location_assignment